### DISCUSSION TITLE
v1.9.0-alpha.0

### DISCUSSION THREAD

**Discussion by markerikson** [reduxjs/redux-toolkit#2620] [Announcements]
This **feature preview** release adds new options and improves behavior for RTK Query, adds a runtime deprecation for the "object" form of `createReducer/createSlice.extraReducers`, adds the ability to define a "pre-typed" version of `createAsyncThunk`, improves TS inference of store enhancers, and exports additional TS types.

We hope to add a couple additional RTKQ options as part of the final 1.9 release, including an `upsertQueryData` util.  See [the RTK 1.9 milestone]() for remaining items.  No hard ETA yet, but ideally we'd like to publish 1.9 within the next couple weeks if we can wrap things up.

## Notable Changes

Some highlights for this alpha:

### RTKQ `merge` Option

RTKQ was built around the assumption that the server is the source of truth, and every refetch replaces the cached data on the client.  There _are_ use cases when it would be useful to merge an incoming response into the existing cached data instead, such as pagination or APIs that return varying results over time.

Query endpoints can now accept a `merge(cachedData, responseData)` callback that lets you do Immer-powered "mutations" to update the existing cached data instead of replacing it entirely.

### Object Reducer Deprecation Warning

RTK's `createReducer` API was originally designed to accept a lookup table of action type strings to case reducers, like `{ "ADD_TODO" : (state, action) => {} }`. We later added [the "builder callback" form]() to allow more flexibility in adding "matchers" and a default handler, and did the same for `createSlice.extraReducers`.

**We intend to remove the "object" form for both `createReducer` and `createSlice.extraReducers` in RTK 2.0**.  The builder callback form is effectively the same number of lines of code, and works _much_ better with TypeScript.

Starting with this release, **RTK will print a one-time runtime warning for both `createReducer` and `createSlice.extraReducers` if you pass in an object argument**.

As an example, this:

```js
const todoAdded = createAction('todos/todoAdded');

createReducer(initialState, {
  [todoAdded]: (state, action) => {}
})

createSlice({
  name,
  initialState,
  reducers: {/* case reducers here */},
  extraReducers: {
    [todoAdded]: (state, action) => {}
  }
})
```

should be migrated to:

```js
createReducer(initialState, builder => {
  builder.addCase(todoAdded, (state, action) => {})
})

createSlice({
  name,
  initialState,
  reducers: {/* case reducers here */},
  extraReducers: builder => {
    builder.addCase(todoAdded, (state, action) => {})
  }
})
```

We have initial codemods in the repo that will help rewrite the object form to the builder form, and we'll publish those with instructions alongside 1.9 when it goes final.

### RTKQ Internal Improvements

When query hooks mount, they dispatch actions to subscribe to the relevant data. The first hook to do so will dispatch a "subscription/fulfilled" action, and all further hooks asking for the same cache key will dispatch "subscription/rejected" actions.  Both cause the reducer logic to add another entry to the subscription tracking.

The dispatching of individual "subscription/rejected" actions was causing perf issues when many components mounted at once, due to the number of extra dispatches. RTKQ now batches those into a single combined action per event loop tick, which improves perf for some many-component use cases noticeably.

### TS Types

`configureStore` now correctly infers changes to the store shape from any store enhancers.

There's now a `createAsyncThunk.withTypes()` method that can be used to create a "pre-typed" version of `createAsyncThunk` with types like `{state, dispatch, extra}` baked in.

## What's Changed

* Add `isJsonContentType` predicate to `fetchBaseQuery` by @msutkowski in https://github.com/reduxjs/redux-toolkit/pull/2331
* Add `jsonContentType` to `fetchBaseQuery` options by @msutkowski in https://github.com/reduxjs/redux-toolkit/pull/2403
* Add ThunkMiddleware to the re-exported types from redux-thunk by @orta in https://github.com/reduxjs/redux-toolkit/pull/2451
* add timeout option to fetchBaseQuery by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/2143
* createSlice: use template literal types for action type by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/2250
* add types for manually typing hook results in userland code by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/2276
* Add 'content-type' ResponseHandler by @taylorkline in https://github.com/reduxjs/redux-toolkit/pull/2363
* feature : endpoint-specific args serializers by @michal-kurz in https://github.com/reduxjs/redux-toolkit/pull/2493
* RFC: add "merge" functionality by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1059
* Add runtime deprecation warning for reducer object notation by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/2591
* feat: Add the ability to type `StoreEnhancers` by @fostyfost in https://github.com/reduxjs/redux-toolkit/pull/2550
* Batch RTKQ "subscribe on reject" actions, and improve cache collection timer handling by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/2599
* add transformErrorReponse to rtkq endpoints by @dreyks in https://github.com/reduxjs/redux-toolkit/pull/1841
* Fix manually initiate()d rtk-query promises by @wesen in https://github.com/reduxjs/redux-toolkit/pull/2187
* Implement codemods for createReducer and createSlice builder by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/2602
* RFC: Expose endpoint types by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1646
* add createAsyncThunk.withTypes by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/2604

**Full Changelog**: https://github.com/reduxjs/redux-toolkit/compare/v1.8.5...v1.9.0-alpha.0

<hr /><em>This discussion was created from the release <a href='https://github.com/reduxjs/redux-toolkit/releases/tag/v1.9.0-alpha.0'>v1.9.0-alpha.0</a>.</em>

**Comment by gordon-eliel** [comment]
Hey @markerikson I am playing with the new `merge` functionality and I'm running into some issues using it for `infinite pagination`. If I pass in a page param, in the api slice, a new cache entry is created per page, which does not execute the `merge` function. I tried using a static cache key for the query but the hook does not get executed, I'm assuming its because the cache key does not change.

Question is how is this feature supposed to be used from a query perspective?

**Reply by markerikson** [reply]
As currently implemented, the `merge` callback will _not_ be run if the existing `cacheEntry.data` value is `undefined` - there's nothing to "merge" into.  (This also sidesteps a question of "should the callback let you 'mutate' or 'return'?").  Instead, if there's no cache data yet, the existing response gets saved as-is, as usual.

In other words, the idea of "merging" only makes sense if there's previously-fetched-and-cached data to merge into.

Is that an issue for what you're trying to do?

The other potential issue here is the query cache key behavior.  RTKQ will still, by default, create an entirely separate cache entry for every single unique set of query args, based on the serialization.  If you're passing in a `page` param, ie `useGetSomeDataQuery({page: 123})`, that creates a different cache key every time you pass in a different page number.

You'll likely want to also use the new `serializeQueryArgs` override option, on a per-endpoint basis, to remove the `page` arg from the cache key generation:

https://github.com/reduxjs/redux-toolkit/pull/2493

That way you have _one_ cache entry for "all the paginated data", changing the `page` value should still trigger requests, but the additional future responses should be able to merge into the one cache entry.

**Reply by gordon-eliel** [reply]
I guess I might be using the api wrong. What happen is, the query if executed the first time with `page=1`, this returns the correct result saved into the store, but when the `page` param changes, the query never gets executed again. I have setup `serializeQueryArgs` to the `endpointName`. This is basically what I have for the query:

```
  const [pageNumber, setPageNumber] = useState(1);
  const {
    data,
    isFetching,
    isLoading,
    refetch,
  } = useListTransactionsQuery(
    { pageNumber },
    {
      pollingInterval: 3000,
      refetchOnMountOrArgChange: true,
      skip: false,
    }
  );
```

And for the endpoint I have for serializing the args:
```
serializeQueryArgs: ({ endpointName }) => {
     return endpointName;
},
```

The query never gets executed when `pageNumber` changes

**Reply by markerikson** [reply]
Hmm. It's certainly possible that we're not re-fetching because the `serializeQueryArgs` result hasn't changed. I wasn't entirely sure what the actual behavior here would be.

Could you put together a CodeSandbox that shows this happening?  This is why we have alphas :)

**Reply by gordon-eliel** [reply]
Here is a sandbox that demonstrates the issues:

https://codesandbox.io/s/merge-query-sandbox-4fmws8?file=/src/Pokemons.tsx

**Reply by gordon-eliel** [reply]
I think this line may be the issue:
https://github.com/reduxjs/redux-toolkit/blob/3de00f4d49e0a7372605531d47bd49f808bd18c2/packages/toolkit/src/query/react/buildHooks.ts#L614

I'll take a look at some possible solutions, this is my first time looking into the codebase so I might be making some assumptions here

**Reply by markerikson** [reply]
Okay, found what's going on.

Made a replay , and if you open up https://app.replay.io/recording/rtk-190-alpha-merge-pagination-not-refetching--13800b41-6680-402f-a564-a80a280a74d7?point=9735556619499725871884967098715939&time=4275&hasFrames=false , you can see that what's happening:

- The query hook _does_ see a new `stableArg` value, and dispatches `initiate()` again
- This _does_ execute `queryThunk`...
- but the `condition` check inside of `queryThunk` ends up retrieving the existing cache entry for the same cache key, sees that it's already in a `"fulfilled"` status, and bails out without trying to fetch the data.

That makes sense, conceptually.  The question is what to _do_ about it.

Here's the current `condition` logic for reference:

```ts
    condition(arg, { getState }) {
      const state = getState()
      const requestState = state[reducerPath]?.queries?.[arg.queryCacheKey]
      const fulfilledVal = requestState?.fulfilledTimeStamp

      // Order of these checks matters.
      // In order for `upsertQueryData` to successfully run while an existing request is in flight,
      /// we have to check for that first, otherwise `queryThunk` will bail out and not run at all.
      if (isUpsertQuery(arg)) return true

      // Don't retry a request that's currently in-flight
      if (requestState?.status === 'pending') return false

      // if this is forced, continue
      if (isForcedQuery(arg, state)) return true

      // Pull from the cache unless we explicitly force refetch or qualify based on time
      if (fulfilledVal)
        // Value is cached and we didn't specify to refresh, skip it.
        return false

      return true
    },
```

Filed https://github.com/reduxjs/redux-toolkit/issues/2684 so we can track this separately - let's continue any discussion over there.
